compiler/rustc_span/src/lev_distance.rs

```diff
@@ -46,6 +46,49 @@ pub fn lev_distance(a: &str, b: &str, limit: usize) -> Option<usize> {
     (dcol[m] <= limit).then_some(dcol[m])
 }
 
+/// Provides a method name similarity score between two method names. The lower the score,
+/// the closer the match. 0 is an identical match.
+///
+/// Uses the Levenshtein distance between the two strings and attempts to account
+/// for cases where one is a substring of the other.
+///
+/// Returns None if the distance exceeds the limit
+pub fn method_name_similarity_score(a: &str, b: &str, limit: usize) -> Option<usize> {
+    let n = a.chars().count();
+    let m = b.chars().count();
+
+    // Check one isn't less than half the length of the other
+    let big_len_diff = (n * 2) < m || (m * 2) < n;
+    let len_diff = if n < m { m - n } else { n - m };
+    let lev = lev_distance(a, b, limit + len_diff)?;
+
+    let score = lev - len_diff;
+
+    let score = {
+        if score == 0 && len_diff > 0 && !big_len_diff {
```

Comment:
Some comments might be helpful to explain the rationale behind these conditionals - they make sense if I think long enough but it's easier when I don't need to :)

Reply:
Added some comments around this and gave a bit more context for what we're trying to do in the rustdoc